### PR TITLE
Integrate CDON clothing data set

### PR DESCRIPTION
**Pull request by alessioGalatolo**
Added CDON dataset support.
Improved overall modularity.
In particular, modularity in FastTensorDataLoader which is now compliant with the semantics of the default python Dataloader
Modularity in the load of CIFAR datasets

### CODE REVIEW

**Comment by rlindsberg** [comment]
Alessio, you have to create new branch from master and rebase your changes. Please don't continue on your own branch.

**Comment by rlindsberg** [comment]
Skærmbillede 2021-05-17 kl  17 18 53


I have now rebased it for you:
Skærmbillede 2021-05-17 kl  17 21 59


Please run in Git:
git fetch
git checkout cdon-dataset-implementation
git reset --hard origin/cdon-dataset-implementation

File: dd2424-google [hunk 1]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN OPENSSH PRIVATE KEY-----
```
Comment (rlindsberg) [inline dd2424-google:1]: This is not good...

**Review by rlindsberg — CHANGES_REQUESTED** [review]

File: dd2424-google (hunk 1 above)
Comment (alessioGalatolo) [reply dd2424-google:1]: Yep, I messed up when committing. Can we undo that commit?

Comment (rlindsberg) [reply dd2424-google:1]: Yes, do:
git fetch
git checkout cdon-dataset-implementation
git reset --hard origin/cdon-dataset-implementation
git reset --soft HEAD~1

**Comment by rlindsberg** [comment]
Don't run away. You can't really delete a commit!

**Comment by rlindsberg** [comment]
I will look at this tonight after some other tasks. Could you provide a bit more detail about what you did? Like a background, methods and results? A few sentences would be nice! Thank you in advance!

**Comment by alessioGalatolo** [comment]
You can see the Custom_dataset file has been adapted to how I've saved the CDON dataset.
We also realized with @YinengWang that the class FastTensorDataLoader does not adapt very well to CDON dataset since this class tries to have in memory the whole dataset and CDON's one is definitely too big. Therefore we adapted said class to be used interchangeably with the PyTorch default DataLoader. I've also merged the functions for loading CIFAR10 and CIFAR100 into one since they were basically identical. I've also found and fixed a bug about the use of memorization parameters when not available and clean some old code (very few things)

**Comment by rlindsberg** [comment]
@alessioGalatolo So I have rewritten your comments, is this correctly understood? Do we remove the FastTensorDataLoader and just use the default one for CIFAR and CDON?

## Background
We also realized with @YinengWang that the class FastTensorDataLoader does not adapt very well to CDON dataset since this class tries to have in memory the whole dataset and CDON's one is definitely too big.

## Solution

- Therefore we adapted said class to be used interchangeably with the PyTorch default DataLoader.
- I've also merged the functions for loading CIFAR10 and CIFAR100 into one since they were basically identical.
- I've also found and fixed a bug about the use of memorization parameters when not available and clean some old code (very few things)

## Results

- The Custom_dataset file has been adapted to how I've saved the CDON dataset
- Improved overall modularity. In particular, modularity in FastTensorDataLoader which is now compliant with the semantics of the default python Dataloader

**Comment by alessioGalatolo** [comment]
Yep, those were the changes. I don't know if it's the case to remove the entire use of FastTensorDataLoader since it will require further changes to how the noise is generated. Maybe @YinengWang can give his opinion

**Comment by rlindsberg** [comment]
Nice!
Also, could you please show some results of your implementation? Liks screenshots of different variables in your IDE and how many seconds / epoch when training different images?

**Comment by YinengWang** [comment]
We don't have to care about adaptions to the noise generator, since now it is n configure option (`config.compute_memorization`) whether noise is considered in training. If the dataset doesn't contain artificial noise, we just disable `config.compute_memorization` and everything goes well.

**Comment by alessioGalatolo** [comment]
> Also, could you please show some results of your implementation? Liks screenshots of different variables in your IDE and how many seconds / epoch when training different images?
Either I don't understand what you mean or they are not relevant for this PR.
Also, I've made this PR a draft since I realized it needs further work in the ELR part

**Comment by rlindsberg** [comment]
> I don't understand what you mean

I meant if you could show some screenshots of your results to help us understand your solution. And show some results of your unit tests (to make sure everything new is working, everything old is not broken).

**Comment by alessioGalatolo** [comment]
Okay, will do once I fix the other issue!

**Comment by rlindsberg** [comment]
We only have 1.5 day left. Can you do it quickly?

**Comment by alessioGalatolo** [comment]
I've already made the fix and I'm testing it but I wanted to confront Alfred to double-check I didn't alter the semantics before pushing anything

**Comment by alessioGalatolo** [comment]
Code should be okay. Will follow graphs of unit test

**Review by rlindsberg — COMMENTED** [review]
Great! Is it ready for review?

**Comment by rlindsberg** [comment]
Still waiting for your screenshots...

**Comment by rlindsberg** [comment]
Sorry for late reply, because Github does not send email notifications when you marked this pull request as ready for review.

**Comment by alessioGalatolo** [comment]
I cannot run the unit tests right now, as I'm still training the whole CDON dataset. I'll post them once it finishes

**Comment by rlindsberg** [comment]
How were the results?

**Comment by alessioGalatolo** [comment]
image
image
Graph of training loss and test accuracy on CIFAR10 with 20% noise of few epochs

**Comment by rlindsberg** [comment]
Thanks! What about the results on CDON images?

**Comment by alessioGalatolo** [comment]
image
This one is with ELR
image
This one is with CE

**Comment by YinengWang** [comment]
https://github.com/YinengWang/Supervised-Image-Classification-with-Noisy-Labels-Using-Deep-Learning/blob/f323e99bb12d4aa32602ac932bdf4f2062aca766/Custom_dataset.py#L23-L26

What does this code do? Isn't `y_wild_label == y_wild_label` always `True`?

**Comment by alessioGalatolo** [comment]
> https://github.com/YinengWang/Supervised-Image-Classification-with-Noisy-Labels-Using-Deep-Learning/blob/f323e99bb12d4aa32602ac932bdf4f2062aca766/Custom_dataset.py#L23-L26
> 
> What does this code do? Isn't `y_wild_label == y_wild_label` always `True`?

It is false if y_wild_label is None. I know the simplest way to check this would be:
`if x is None:`
but it gave me problem when y_wild_label was not none and was a numpy array. I would get something like "check on arrays is ambiguous"

**Comment by alessioGalatolo** [comment]
With @Alfred-N we tested that the distribution of the categories was uniform, here is a plot proving this:
test_labels
x-axis is the subcategory number, y axis is the number of samples. orange is the test set, blue is the training

**Comment by Alfred-N** [comment]
> With @Alfred-N we tested that the distribution of the categories was uniform, here is a plot proving this:
> test_labels
> x-axis is the subcategory number, y axis is the number of samples. orange is the test set, blue is the training

Yes, so there was a bug with the distribution of the test set, but as we can see from the plot it's fixed now.

**Review by Alfred-N — DISMISSED** [review]
Great work

**Comment by rlindsberg** [comment]
> It is false if y_wild_label is None. I know the simplest way to check this would be:
> `if x is None:`
> but it gave me problem when y_wild_label was not none and was a numpy array. I would get something like "check on arrays is ambiguous"

Skærmbillede 2021-05-19 kl  19 36 06

Also, according to PEP8, we should use:
if xxx is not None:
   do something

https://www.python.org/dev/peps/pep-0008/#programming-recommendations

**Comment by rlindsberg** [comment]
Skærmbillede 2021-05-19 kl  19 40 54
> but it gave me problem when y_wild_label was not none and was a numpy array. I would get something like "check on arrays is ambiguous"

@alessioGalatolo Is this what you referred to?

**Comment by Alfred-N** [comment]
> > It is false if y_wild_label is None. I know the simplest way to check this would be:
> > `if x is None:`
> > but it gave me problem when y_wild_label was not none and was a numpy array. I would get something like "check on arrays is ambiguous"
> 
> Skærmbillede 2021-05-19 kl 19 36 06
> 
> Also, according to PEP8, we should use:
> if xxx is not None:
> do something
> 
> https://www.python.org/dev/peps/pep-0008/#programming-recommendations

Does this matter? or is it just aesthetic

**Comment by rlindsberg** [comment]
> > > It is false if y_wild_label is None. I know the simplest way to check this would be:
> > > `if x is None:`
> > > but it gave me problem when y_wild_label was not none and was a numpy array. I would get something like "check on arrays is ambiguous"
> > 
> > 
> > Skærmbillede 2021-05-19 kl 19 36 06
> > Also, according to PEP8, we should use:
> > if xxx is not None:
> > do something
> > https://www.python.org/dev/peps/pep-0008/#programming-recommendations
> 
> Does this matter? or is it just aesthetic

Yes, because that if-statement will always be true. And the program will crash if the variable is a numpy array

**Comment by alessioGalatolo** [comment]
> Skærmbillede 2021-05-19 kl 19 40 54
> 
> > but it gave me problem when y_wild_label was not none and was a numpy array. I would get something like "check on arrays is ambiguous"
> 
> @alessioGalatolo Is this what you referred to?

Yep, the error is that one though I'm not sure the object was a numpy array as by logic we should get a slice of pandas.DataFrame


> > It is false if y_wild_label is None. I know the simplest way to check this would be:
> > `if x is None:`
> > but it gave me problem when y_wild_label was not none and was a numpy array. I would get something like "check on arrays is ambiguous"
> 
> Skærmbillede 2021-05-19 kl 19 36 06
> 
> Also, according to PEP8, we should use:
> if xxx is not None:
> do something
> 
> https://www.python.org/dev/peps/pep-0008/#programming-recommendations

The result of object == object depends also on the object as the equals function can be overwritten.
In the case I was facing this was the only solution that worked.

**Comment by rlindsberg** [comment]
Can you describe what happens if you use:
if xxx is not None?

**Comment by alessioGalatolo** [comment]
Anyway, that line was used in the previous dataset file in which some values were missing. That line made the program robust to such problems.
However, the dataset I'm using now does not have missing values. So if that statement really bothers you so much you can remove it and I _belive_ it will keep working

**Comment by rlindsberg** [comment]
Skærmbillede 2021-05-19 kl  20 53 58

This, however, will work.

**Review by rlindsberg — DISMISSED** [review]

**Comment by YinengWang** [comment]
https://github.com/YinengWang/Supervised-Image-Classification-with-Noisy-Labels-Using-Deep-Learning/blob/92e7363b2b8fe22132d9bcc5326d8b32873fe43c/datasets.py#L94-L95

We are using FastTensorDataLoader for the CDON dataset. For small amount of data it's okay, but it has potential memory problems.

> We also realized with @YinengWang that the class FastTensorDataLoader does not adapt very well to CDON dataset since this class tries to have in memory the whole dataset and CDON's one is definitely too big.

We can still use PyTorch's DataLoader, this won't causes trouble because we have [added a config option](https://github.com/YinengWang/Supervised-Image-Classification-with-Noisy-Labels-Using-Deep-Learning/pull/18#issuecomment-843058832).

**Comment by rlindsberg** [comment]
I believe PyTorch's DataLoader will use less CPU power also. For server2, 1.5 cores vs 4 cores.

**Review by rlindsberg — APPROVED** [review]

**Comment by rlindsberg** [comment]
@alessioGalatolo 😍 Nice!
